core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -0,0 +1,116 @@
+import org.apache.kafka.common.{Node, TopicPartition}
+
+import scala.collection.{concurrent, immutable, mutable}
+import collection.JavaConverters._
+
+class TransactionMarkerChannel(interBrokerListenerName: ListenerName, metadataCache: MetadataCache) extends Logging {
+
+  // we need the metadataPartition so we can clean up when Transaction Log partitions emigrate
+  case class PendingTxnKey(metadataPartition: Int, producerId: Long)
+
+
+  private[transaction] val brokerStateMap: concurrent.Map[Int, DestinationBrokerAndQueuedMarkers] = concurrent.TrieMap.empty[Int, DestinationBrokerAndQueuedMarkers]
+  private val pendingTxnMap: concurrent.Map[PendingTxnKey, TransactionMetadata] = concurrent.TrieMap.empty[PendingTxnKey, TransactionMetadata]
+
+  def addOrUpdateBroker(broker: Node) {
+    if (brokerStateMap.contains(broker.id())) {
+      val brokerQueue = brokerStateMap(broker.id())
+      if (!brokerQueue.destBrokerNode.equals(broker)) {
+        brokerStateMap.put(broker.id(), DestinationBrokerAndQueuedMarkers(broker, brokerQueue.markersQueue))
+        trace(s"Updated destination broker for ${broker.id} from: ${brokerQueue.destBrokerNode} to: $broker")
+      }
+    } else {
+      val markersQueue = new LinkedBlockingQueue[CoordinatorEpochAndMarkers]()
+      brokerStateMap.put(broker.id, DestinationBrokerAndQueuedMarkers(broker, markersQueue))
+      trace(s"Added destination broker ${broker.id}: $broker")
+    }
+  }
+
+  private[transaction] def addRequestForBroker(brokerId: Int, txnMarkerRequest: CoordinatorEpochAndMarkers) {
+    val markersQueue = brokerStateMap(brokerId).markersQueue
+    markersQueue.add(txnMarkerRequest)
+    trace(s"Added markers $txnMarkerRequest for broker $brokerId")
+  }
+
+  def addRequestToSend(metadataPartition: Int, pid: Long, epoch: Short, result: TransactionResult, coordinatorEpoch: Int, topicPartitions: immutable.Set[TopicPartition]): Unit = {
+    val partitionsByDestination: immutable.Map[Int, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
+      val leaderForPartition = metadataCache.getPartitionInfo(topicPartition.topic, topicPartition.partition)
+      val currentBrokers = mutable.Set.empty[Int]
+      leaderForPartition match {
+        case Some(partitionInfo) =>
+          val brokerId = partitionInfo.leaderIsrAndControllerEpoch.leaderAndIsr.leader
+          if (currentBrokers.add(brokerId)) {
+            // TODO: What should we do if we get BrokerEndPointNotAvailableException?
+            val broker = metadataCache.getAliveEndpoint(brokerId, interBrokerListenerName).get
+            addOrUpdateBroker(broker)
+          }
+          brokerId
+        case None =>
+          // TODO: there is a rare case that the producer gets the partition info from another broker who has the newer information of the
+          // partition, while the TC itself has not received the propagated metadata update; do we need to block when this happens?
```

Comment:
Hmm, instead of throwing IllegalStateException, it seems that we should just keep retrying until successful?